1 scoping

lizgzil opened the pull request:
---

# Description

This is a refactor of the work Adeola did in training the model, but this time using batch.

The full flow can be run by running:
```
cd asf_floorplan_interpreter/pipeline/
python train_yolo.py --package-suffixes=.txt,.yaml,.jpg --datastore=s3 run --config_file window_door_config.yaml
```

or the smaller test version:
```
cd asf_floorplan_interpreter/pipeline/
python train_yolo.py --package-suffixes=.txt,.yaml,.jpg --datastore=s3 run --config_file window_door_test_config.yaml
```


This will train the model and output the best model in [this S3 location](https://s3.console.aws.amazon.com/s3/buckets/asf-floorplan-interpreter?region=eu-west-2&prefix=window_door_config/&showversions=false).

The metaflow files are in this location: https://s3.console.aws.amazon.com/s3/buckets/open-jobs-lake?prefix=metaflow/FloorPlanYolo/&region=eu-west-1


It is super quick too! took about 5-10 minutes to train.

Still lots to do - e.g. add newly labelled data, make sure the validation is outputting properly? But I think thats for another PR!

Fixes #5 

# Instructions for Reviewer

Try out 
```
export METAFLOW_PROFILE=floorplan
cd asf_floorplan_interpreter/pipeline/
python train_yolo.py --package-suffixes=.txt,.yaml,.jpg --datastore=s3 run --config_file configs/window_door_test_config.yaml
```
you will need to create the correct metaflow config (and save it in `~/.metaflowconfig/config_floorplan.json`), which I can send you.

# Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [ ] I have checked the code runs
- [ ] I have tested the code
- [ ] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained this PR above
- [ ] I have requested a code review

dede95 committed:
Add scripts for training

dede95 committed:
Changed function for local use

dede95 committed:
Files for prediction

lizgzil committed:
Bath metaflow experimentation

lizgzil commented:
I had to manually add the metaflow env vars e.g. "METAFLOW_BATCH_JOB_QUEUE".

There are issues with the conda decorator finding `ultralytics` (even though it is [here](https://anaconda.org/conda-forge/ultralytics)), so I couldnt get `from ultralytics import YOLO` to work at all in batch.

Even commenting out the ultralytics bits so that the train step is just `results=2` I got it to work once, but then it failed with:
```
2023-10-10 17:02:43.759 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792] Task is starting.
2023-10-10 17:02:44.378 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792] Bootstrapping environment...
2023-10-10 17:02:46.468 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792] Traceback (most recent call last):
2023-10-10 17:02:46.468 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792]   File "/usr/local/lib/python3.8/runpy.py", line 194, in _run_module_as_main
2023-10-10 17:02:46.468 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792]     return _run_code(code, main_globals, None,
2023-10-10 17:02:46.468 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792]   File "/usr/local/lib/python3.8/runpy.py", line 87, in _run_code
2023-10-10 17:02:46.468 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792]     exec(code, run_globals)
2023-10-10 17:02:46.468 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792]   File "/metaflow/metaflow/plugins/conda/batch_bootstrap.py", line 104, in <module>
2023-10-10 17:02:46.468 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792]     bootstrap_environment(sys.argv[1], sys.argv[2], sys.argv[3])
2023-10-10 17:02:46.468 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792]   File "/metaflow/metaflow/plugins/conda/batch_bootstrap.py", line 16, in bootstrap_environment
2023-10-10 17:02:46.468 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792]     packages = download_conda_packages(flow_name, env_id, datastore_type)
2023-10-10 17:02:46.469 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792]   File "/metaflow/metaflow/plugins/conda/batch_bootstrap.py", line 75, in download_conda_packages
2023-10-10 17:02:46.469 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792]     shutil.move(
2023-10-10 17:02:46.469 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792]   File "/usr/local/lib/python3.8/shutil.py", line 791, in move
2023-10-10 17:02:46.469 [1696957156421451/train/3 (pid 16813)] [a59e97ec-aa3a-4426-ad2d-65236f6a7792]     os.rename(src, real_dst)
2023-10-10 17:02:49.781 [1696957156421451/train/3 (pid 16813)] Data store error:
2023-10-10 17:02:49.782 [1696957156421451/train/3 (pid 16813)] No completed attempts of the task was found for task 'FloorPlanYolo/1696957156421451/train/3'
2023-10-10 17:02:49.863 [1696957156421451/train/3 (pid 16813)]
```

Jack-Vines committed:
Add custom reqs/config and update flow

lizgzil committed:
download data for metaflow

lizgzil committed:
Correct batch flow and create config for full run

lizgzil committed:
reset output locations for model

lizgzil committed:
clean up flow

lizgzil committed:
remove redundant configs

lizgzil commented on requirements.txt:
```diff
@@ -0,0 +1,6 @@
+numpy
+pandas
+ultralytics
+sklearn
```
```suggestion
scikit-learn
```

lizgzil committed:
Correct sklearn

lizgzil committed:
Finalise processing prodigy data for training, and tweaks to training script, update readmes

lizgzil committed:
Add new configs and put them all in a subfolder

lizgzil commented:
Also includes edits to the `prodigy_to_yolo.py` script.

As of the 191023 dataset we have:

#### rooms dataset
Original data was 122 annotations, filtered for accepted labels and deduplicated gives us 105 annotations
- Saving 63 image in the training set
- Saving 21 image in the test set
- Saving 21 image in the val set

#### window/door/staircase dataset
Original data was 119 annotations, filtered for accepted labels and deduplicated gives us 102 annotations
-  Saving 61 image in the training set
- Saving 20 image in the test set
- Saving 21 image in the val set

cmbrennan002 commented on README.md:
```diff
@@ -7,6 +7,8 @@
   - Setup the conda environment
   - Configure `pre-commit`
+  - Set-up Metaflow `export METAFLOW_PROFILE=floorplan` and `direnv reload`
+  - Make sure your `METAFLOW_USER` name has no fullstops in.
```
Love the pre-emptive trouble shooting :)

cmbrennan002 commented on asf_floorplan_interpreter/pipeline/README.md:
```diff
@@ -0,0 +1,63 @@
+
+Note that for Batch to work, the configs need to be in the same folder as `train_yolo.py` - hence why we store them in the `configs/` folder rather than elsewhere in this repo.
+
+To train each model a pair of configs are needed; the main config (e.g. `room_config.yaml`) and the config neccessary for training using YOLO (`yolo_room_config.yaml`). The latter is referenced in the former, so when you run `train_yolo.py` you only need the first config in the argument.
```
I am getting an error when I run the model that says 'No such file or directory: 'configs/window_door_test_config.yaml', and on VS Studio code the config files seem to be in the same folder as the train_yolo.py (as you say). I ammended the command to remove the 'configs' and it worked (so would be python train_yolo.py --package-suffixes=.txt,.yaml,.jpg --datastore=s3 run --config_file CONFIG_NAME.yaml)

lizgzil replied:
I think this was fixed now?

cmbrennan002 commented on asf_floorplan_interpreter/pipeline/configs/yolo_roboflow_config.yaml:
```diff
@@ -0,0 +1,9 @@
+save_dir: "roboflow_save_dir/"
+
+nc: 6
+names: ["DOOR", "DOUBLE DOOR", "FOLDING DOOR", "ROOM", "SLIDING DOOR", "WINDOW"]
```
Should we explain somewhere how the 'nc' works? Each class is mapped to a number. We have manually re-coded the roboflow original classes (shown here) to collapse categories of doors and windows, and split out rooms.

lizgzil replied:
good point - I've added info to the readme

cmbrennan002 commented on asf_floorplan_interpreter/pipeline/flow_reqs.txt:
```diff
@@ -0,0 +1,6 @@
+numpy
+pandas
+ultralytics
```
Should we also specify a version for this? I feel like everything else is pretty standard but I don't know how the versions work for this package. Mine is 8.0.191

lizgzil replied:
yep! will do. Mine is 8.0.196 though, maybe I can do `>=8.0.191`

cmbrennan002 commented on asf_floorplan_interpreter/pipeline/predict_floorplan.py:
```diff
@@ -0,0 +1,28 @@
+
+def predict_image(directory, img):
+    "Collects model and applies to chosen images"
+    model = load_model(
```
This code looks good, and not substantially changed from what we had.

This might be me just not understanding how to use GIthub/ S3 correctly, but I read this function as needing to put in the S3 model path as the directory (unless I should have already loaded it in a previous step to my local directory)? But then the PROJECT_DIR is the GIthub route folder, so the directory becomes garbled. It might be helpful to add some instructions (maybe in the getters)? Or if this is just a standard way of working I don't know yet, let me know!

lizgzil replied:
yes - it's actually Adeola's original code!

I think but I'm not 100% sure that we need to load the model from a local location rather than S3. Either way - as you point out, this is what the function is assuming by using PROJECT_DIR - so I will add that to the docstring and variable name just so it's cleaner. 

Perhaps later we could look into whether its easy to load directly from S3 (I had a little go but it wasn't that easy)

cmbrennan002 commented on asf_floorplan_interpreter/pipeline/prodigy_to_yolo.py:
```diff
@@ -48,43 +54,117 @@ def convert_prod_to_yolo(prod_label, object_to_class_dict):
+    # Original data was 119 annotations, filtered for accepted labels and deduplicated gives us 102 annotations
+    # Saving 61 image in the training set
+    # Saving 20 image in the test set
+    # Saving 21 image in the val set
```
This all looks good, but I haven't tested it, because I don't want to overwrite labelled files. Maybe we can talk through quickly next week.

cmbrennan002 commented on asf_floorplan_interpreter/pipeline/train_yolo.py:
```diff
@@ -0,0 +1,97 @@
+
+
+if __name__ == "__main__":
+    FloorPlanYolo()
```
Looks good, and once the 'config' thing was changed, ran perfectly on my machine!

cmbrennan002 commented on asf_floorplan_interpreter/pipeline/train_yolo.py:
```diff
@@ -0,0 +1,97 @@
+import os
+
+try:
+    os.system("apt-get update && apt-get install -y libgl1-mesa-glx 1> /dev/null")
```
I always get an error that this doesn't work. It's not an issue because of the 'try' statement, but is it worth a line of explanation for what this does/ who would need it to work?

lizgzil replied:
yes - good point! It will always error, but its needed for the batch step. Will add a comment

cmbrennan002 commented on asf_floorplan_interpreter/utils/model_utils.py:
```diff
@@ -0,0 +1,13 @@
+from ultralytics import YOLO
+
+
+def load_model(directory):
```
I think this is not yet edited/ superseded by the getter?

lizgzil replied:
yes - it's code from Adeola. It should still work though, but its only used in `predict_floorplan.py`. Will leave for a future PR when we use the model more then!

lizgzil committed:
Addressing Clare's PR comments

lizgzil commented:
@cmbrennan002 I've addressed your comments in https://github.com/nestauk/asf_floorplan_interpreter/pull/4/commits/200d5113d40c31f85348e4ca20df84b6ae4ee543 🎉  - thanks for the review!

Am I ok to merge now?
